src/v/cloud_storage/partition_manifest.cc

```diff
@@ -860,6 +884,121 @@ partition_manifest::segment_containing(model::offset o) const {
+
+    // Fast handling of bounds/edge cases to simplify subsequent
+    // arithmetic steps
+    if (t < base_t) {
+        return _segments.begin()->second;
+    } else if (t > max_t) {
+        return std::nullopt;
+    } else if (max_t == base_t) {
+        // This is plausible in the case of a single segment with
+        // a single batch using LogAppendTime.
+        return _segments.begin()->second;
+    }
+
+    // Single-offset case should have hit max_t==base_t above
+    vassert(
+      max_offset > base_offset,
+      "Unexpected offsets {} {} (times {} {})",
+      base_offset,
+      max_offset,
+      base_t,
+      max_t);
+
+    // From this point on, we have finite positive differences between
+    // base and max for both offset and time.
+    model::offset interpolated_offset
+      = base_offset
+        + model::offset{
+          (float(t() - base_t()) / float(max_t() - base_t()))
+          * float(max_offset() - base_offset())};
+
+    vlog(
+      cst_log.debug, "timequery t={} offset guess {}", t, interpolated_offset);
+
+    // 2. Convert offset guess into segment guess.  This is not a strictly
+    // correct offset lookup, we just want something close.
+    auto segment_iter = _segments.lower_bound(
+      key{interpolated_offset, model::term_id{-1}});
+    if (segment_iter == _segments.end()) {
+        segment_iter = --_segments.end();
+    }
+    vlog(
+      cst_log.debug,
+      "timequery t={} segment initial guess {}",
+      t,
+      segment_iter->second);
+
+    // 3. Do linear search backward or forward from the location
+    //    we probed, to find the right segment.
+    if (segment_iter->second.base_timestamp < t) {
+        // Our guess's base_timestamp is before the search point, so our
+        // result must be this segment or a later one: search forward
+        for (; segment_iter != _segments.end(); ++segment_iter) {
+            auto base_timestamp = segment_iter->second.base_timestamp;
+            auto max_timestamp = segment_iter->second.max_timestamp;
+            if (max_timestamp >= t) {
+                // We found a segment bounding the search point
+                break;
+            } else if (base_timestamp > t) {
+                // We found the first segment after the search point
+                break;
+            }
+        }
+
+        return segment_iter->second;
+    } else {
+        // Search backwards: we must peek at each preceding segment
+        // to see if its max_timestamp is >= the query t, before
+        // deciding whether to walk back to it.
+        auto segment_riter = std::make_reverse_iterator(segment_iter);
+        vassert(
+          segment_riter->second.base_offset
+          == segment_iter->second.base_offset);
+        if (segment_riter != _segments.rbegin()) {
+            // make_reverse_iterator on a btree map appears to jump one
+            // element backward in the non-reversed order.
+            segment_riter--;
+        }
```

Comment:
> appears to jump one

is it supposed to do that? getting that right seems important for correctness since next we'll only move in one direction

Reply:
I can't find clear docs one way other other.  I started out thinking that `make_reverse_iterator` should point to the same element as the original iterator, but that turned out to be untrue in practice :-/

However... this was already a bidirectional iterator, so I've updated the code to avoid the explicit reverse.